File: desktopclient/src/main/scala/com/goexp/galgame/gui/view/game/listview/tableview/TableController.java
package com.goexp.galgame.gui.view.game.listview.tableview;

import com.goexp.common.util.DateUtil;
import com.goexp.galgame.common.model.GameState;
import com.goexp.galgame.gui.model.Game;
import com.goexp.galgame.gui.task.game.ChangeGameTask;
import com.goexp.galgame.gui.util.*;
import com.goexp.galgame.gui.view.common.jump.JumpBrandController;
import com.goexp.galgame.gui.view.common.jump.JumpLinkController;
import com.goexp.galgame.gui.view.game.HomeController;
import javafx.beans.property.SimpleStringProperty;
import javafx.concurrent.Service;
import javafx.fxml.FXML;
import javafx.scene.control.*;
import javafx.scene.control.cell.PropertyValueFactory;
import javafx.scene.image.ImageView;
import javafx.scene.input.MouseButton;
import javafx.scene.layout.HBox;
import javafx.scene.layout.Region;

import java.time.LocalDate;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class TableController extends DefaultController {

    @FXML
    public TableView<Game> table;
    @FXML
    public TableColumn<Game, List<String>> tableColTag;
    @FXML
    public TableColumn<Game, String> tableColType;
    @FXML
    public TableColumn<Game, String> tableColBrand;
    @FXML
    public TableColumn<Game, String> tableColPainter;
    @FXML
    public TableColumn<Game, String> tableColWriter;
    @FXML
    public TableColumn<Game, GameState> tableColState;
    @FXML
    public TableColumn<Game, Integer> tableColStar;
    @FXML
    public TableColumn<Game, LocalDate> tableColDate;
    @FXML
    public TableColumn<Game, String> tableColTitle;
    @FXML
    public TableColumn<Game, String> tableColCommand;
    @FXML
    private ContextMenu menuPopup;
    private List<Game> selectedGames;


    private Service<Void> changeGameService = new TaskService<>(() -> new ChangeGameTask.MultiLike(selectedGames));


    protected void initialize() {

        var items = Stream.of(GameState.values())
//                .filter(value -> value != GameState.UNCHECKED)
                .sorted(Comparator.reverseOrder())
                .map(brandType -> {
                    var menuItem = new MenuItem();
                    menuItem.setUserData(brandType);
                    menuItem.setText(brandType.getName());
                    menuItem.setOnAction((e) -> {

                        var type = (GameState) (((MenuItem) (e.getSource())).getUserData());
                        logger.debug("<MenuItem>:{}", type.getName());


                        selectedGames = table.getSelectionModel().getSelectedItems();

                        selectedGames.forEach(game -> game.state.set(type));

                        changeGameService.restart();
                    });

                    return menuItem;
                })
                .collect(Collectors.toList());

        menuPopup.getItems().setAll(items);


        table.getSelectionModel().setSelectionMode(SelectionMode.MULTIPLE);

//        table.setOnMouseClicked(event -> {
//            if (event.getButton() == MouseButton.PRIMARY && event.getClickCount() == 2) {
//                var g=table.getSelectionModel().getSelectedItem();
//
//                if(g!=null)
//                {
//                    MainController.$this.loadDetail(g);
//                }
//            }
//
//        });


//        table.addEventHandler(MouseEvent.MOUSE_CLICKED, event -> {
//            if (event.getButton() == MouseButton.PRIMARY && event.getClickCount() == 2) {
//                {
//                    if (event.getTarget() instanceof TableRow) {
//                        var row = (TableRow) event.getTarget();
//
//                        if (!row.isEmpty()) {
//                            var g = (Game) row.getItem();
//                            MainController.$this.loadDetail(g);
//                        }
//                    }
//                }
//            }
//        });


        table.setRowFactory(tr -> {
            var row = new TableRow<Game>();
            row.setOnMouseClicked(null);

            row.setOnMouseClicked(event -> {

                if (event.getButton() == MouseButton.PRIMARY && event.getClickCount() == 2) {
                    if (!row.isEmpty()) {
                        var g = row.getItem();

                        HomeController.$this.loadDetail(g);
                    }
                }
            });

            return row;
        });


        tableColState.setCellValueFactory(new PropertyValueFactory<>("state"));

        tableColBrand.setCellValueFactory(param -> new SimpleStringProperty(param.getValue().brand.name));

        tableColStar.setCellValueFactory(new PropertyValueFactory<>("star"));
        tableColPainter.setCellValueFactory(new PropertyValueFactory<>("painter"));
        tableColWriter.setCellValueFactory(new PropertyValueFactory<>("writer"));
        tableColDate.setCellValueFactory(new PropertyValueFactory<>("publishDate"));
        tableColTag.setCellValueFactory(new PropertyValueFactory<>("tag"));
        tableColTitle.setCellValueFactory(new PropertyValueFactory<>("name"));
        tableColType.setCellValueFactory(new PropertyValueFactory<>("type"));


        tableColDate.setCellFactory(col -> new TableCell<>() {
            @Override
            protected void updateItem(LocalDate item, boolean empty) {
                super.updateItem(item, empty);
                this.setGraphic(null);
                this.setText(null);

                if (item != null && !empty) {
                    this.setText(DateUtil.formatDate(item));
                }
            }
        });

        tableColTag.setCellFactory(col -> new TableCell<>() {
            @Override
            protected void updateItem(List<String> item, boolean empty) {
                super.updateItem(item, empty);
                this.setGraphic(null);
                this.setText(null);

                if (item != null && !empty) {

                    if (item.size() > 0) {
                        var hbox = new HBox();
                        hbox.setSpacing(5);
                        hbox.getChildren().setAll(Tags.toNodes(item));
                        this.setGraphic(hbox);
                    }
                }
            }
        });


        tableColBrand.setCellFactory(col -> {
            final var loader = new FXMLLoaderProxy<Region, JumpBrandController>("view/brand/brandjump.fxml");

            return new TableCell<>() {
                @Override
                protected void updateItem(String item, boolean empty) {
                    super.updateItem(item, empty);
                    this.setGraphic(null);
                    this.setText(null);

                    if (item != null && !empty) {

                        var game = this.getTableRow().getItem();

                        if (game != null) {
                            loader.controller.load(game.brand);

                            this.setGraphic(loader.node);
                        }

                    }
                }
            };
        });

        tableColTitle.setCellFactory(col -> new TableCell<>() {
            @Override
            protected void updateItem(String item, boolean empty) {
                super.updateItem(item, empty);
                this.setGraphic(null);
                this.setText(null);

                if (item != null && !empty) {
                    var game = this.getTableRow().getItem();

                    if (game != null) {
                        var title = game.name;
                        var titleLabel = title.replaceAll("＜[^＞]*＞", "");

                        this.setText(titleLabel);
                    }


                }
            }
        });

        tableColCommand.setCellFactory(col -> {
            final var loader = new FXMLLoaderProxy<Region, JumpLinkController>("view/jump/websitejump.fxml");

            return new TableCell<>() {
                protected void updateItem(String item, boolean empty) {
                    super.updateItem(item, empty);
                    this.setGraphic(null);
                    this.setText(null);

                    if (item != null && !empty) {

                        var game = this.getTableRow().getItem();

                        if (game != null) {
                            Hyperlink viewLink = new Hyperlink("View");
                            viewLink.setOnAction((e) -> {

                                HomeController.$this.loadDetail(game);
                            });
                            loader.controller.load(game);

                            this.setGraphic(new HBox(viewLink, loader.node));
                        }
                    }
                }
            };
        });

        tableColState.setCellFactory(col -> new TableCell<>() {
            @Override
            protected void updateItem(GameState item, boolean empty) {
                super.updateItem(item, empty);
                this.setGraphic(null);
                this.setText(null);

                this.getTableRow().getStyleClass().remove("gray");
                if (item != null && !empty) {
                    if (item == GameState.BLOCK) {
                        this.getTableRow().getStyleClass().add("gray");
                    }
                    this.setText(item.getName());
                }
            }
        });

        tableColStar.setCellFactory(col -> new TableCell<>() {
            @Override
            protected void updateItem(Integer item, boolean empty) {
                super.updateItem(item, empty);
                this.setGraphic(null);
                this.setText(null);

                if (!empty) {
                    var image = LocalRes.HEART_16_PNG.get();
                    var box = new HBox();
                    for (var i = 0; i < item; i++) {
                        box.getChildren().add(new ImageView(image));
                    }

                    this.setGraphic(box);
                }
            }
        });
    }
}

File: desktopclient/src/main/scala/com/goexp/galgame/gui/view/game/listview/tableview/TableController.scala
package com.goexp.galgame.gui.view.game.listview.tableview

import java.time.LocalDate
import java.util

import com.goexp.common.util.date.DateUtil
import com.goexp.galgame.common.model.game.{GameLocation, GameState}
import com.goexp.galgame.gui.model.{Brand, Game}
import com.goexp.galgame.gui.task.game.change.{MultiBlock, MultiLocation, MultiState}
import com.goexp.galgame.gui.view.MainController
import com.goexp.galgame.gui.view.common.control.StarRatingView
import com.goexp.ui.javafx.{DefaultController, TaskService}
import com.goexp.ui.javafx.control.cell.{NodeTableCell, TableCell, TextTableCell}
import javafx.beans.property.SimpleObjectProperty
import javafx.fxml.FXML
import javafx.scene.control._
import javafx.scene.input.MouseButton

import scala.jdk.CollectionConverters._

class TableController extends DefaultController {
  @FXML var table: TableView[Game] = _

  @FXML var tableColDate: TableColumn[Game, LocalDate] = _
  @FXML var tableColStar: TableColumn[Game, Int] = _
  @FXML var tableColBrand: TableColumn[Game, Brand] = _
  @FXML var tableColTitle: TableColumn[Game, String] = _
  @FXML var tableColType: TableColumn[Game, String] = _
  @FXML var tableColPainter: TableColumn[Game, String] = _
  @FXML var tableColWriter: TableColumn[Game, String] = _
  @FXML var tableColState: TableColumn[Game, GameState] = _
  @FXML var tableColLocation: TableColumn[Game, GameLocation] = _

  @FXML private var menuState: Menu = _
  @FXML private var menuLocation: Menu = _

  private var selectedGames: util.List[Game] = _
  final private val changeGameStateService = TaskService(new MultiState(selectedGames))
  final private val blockGameService = TaskService(new MultiBlock(selectedGames))
  final private val changeGameLocationService = TaskService(new MultiLocation(selectedGames))

  override protected def initialize() = {

    def initMenuState() = {
      val items =
        GameState.values.to(LazyList).reverse
          .map(state => {
            val menuItem = new MenuItem
            menuItem.setUserData(state)
            menuItem.setText(state.name)
            menuItem.setOnAction(e => {
              val gameState = ((e.getSource).asInstanceOf[MenuItem]).getUserData.asInstanceOf[GameState]

              logger.debug(s"<MenuItem>:${gameState.name}")

              selectedGames = table.getSelectionModel.getSelectedItems
              selectedGames.forEach(_.state.set(gameState))

              if (gameState == GameState.BLOCK) {

                selectedGames.forEach { g =>
                  g.star = 0
                  g.location.set(GameLocation.REMOTE)
                }

                blockGameService.restart()
              } else {
                changeGameStateService.restart()
              }
            })
            menuItem
          }).asJava

      menuState.getItems.setAll(items)
    }

    def initMenuLocation() = {
      val items =
        GameLocation.values.to(LazyList).reverse
          .map(loc => {
            val menuItem = new MenuItem
            menuItem.setUserData(loc)
            menuItem.setText(loc.name)
            menuItem.setOnAction(e => {
              val gameLoc = ((e.getSource).asInstanceOf[MenuItem]).getUserData.asInstanceOf[GameLocation]

              logger.debug(s"<MenuItem>:${gameLoc.name}")

              selectedGames = table.getSelectionModel.getSelectedItems
              selectedGames.forEach(_.location.set(gameLoc))
              changeGameLocationService.restart()
            })
            menuItem
          }).asJava

      menuLocation.getItems.setAll(items)
    }


    initMenuState()
    initMenuLocation()


    table.getSelectionModel.setSelectionMode(SelectionMode.MULTIPLE)
    table.setRowFactory(_ => {
      val row = new TableRow[Game]
      row.setOnMouseClicked(null)
      row.setOnMouseClicked(event => {
        if ((event.getButton eq MouseButton.PRIMARY) && event.getClickCount == 2)
          if (!row.isEmpty) {
            val g = row.getItem
            MainController().loadDetail(g)
          }
      })
      row
    })
    tableColLocation.setCellValueFactory(param => param.getValue.location)
    tableColState.setCellValueFactory(param => param.getValue.state)
    tableColBrand.setCellValueFactory(param => new SimpleObjectProperty(param.getValue.brand))
    tableColStar.setCellValueFactory(param => new SimpleObjectProperty(param.getValue.star))
    tableColPainter.setCellValueFactory(param => new SimpleObjectProperty(param.getValue.getPainter))
    tableColWriter.setCellValueFactory(param => new SimpleObjectProperty(param.getValue.getWriter))
    tableColDate.setCellValueFactory(param => new SimpleObjectProperty(param.getValue.publishDate))
    tableColTitle.setCellValueFactory(param => new SimpleObjectProperty(param.getValue.name))
    tableColType.setCellValueFactory(param => new SimpleObjectProperty(param.getValue.getType))

    tableColDate.setCellFactory { _ =>
      TextTableCell { publishDate =>
        DateUtil.formatDate(publishDate)
      }
    }

    tableColBrand.setCellFactory(_ =>
      TextTableCell { brand =>
        brand.name
      }
    )

    tableColTitle.setCellFactory(_ =>
      TextTableCell { name =>
        name.replaceAll("＜[^＞]*＞", "")
      }
    )

    tableColLocation.setCellFactory(_ =>
      TextTableCell { location =>
        location.name
      })

    tableColState.setCellFactory(_ => new TableCell[Game, GameState]() {

      override protected def notEmpty(gameState: GameState): Unit = {
        if (gameState eq GameState.BLOCK)
          this.getTableRow.getStyleClass.add("gray")

        this.setText(gameState.name)
      }


      override protected def updateItem(item: GameState, empty: Boolean) = {

        this.getTableRow.getStyleClass.remove("gray")
        super.updateItem(item, empty)

      }
    })


    tableColStar.setCellFactory(_ =>
      NodeTableCell { star =>
        new StarRatingView(star)
      }
    )
  }
}